autoblocks/_impl/tracer.py

```diff
@@ -1,24 +1,55 @@
 import asyncio
+import atexit
+import contextvars
+import dataclasses
+import inspect
 import logging
+import time
 import uuid
 from contextlib import contextmanager
-from dataclasses import dataclass
 from datetime import datetime
 from datetime import timedelta
 from datetime import timezone
 from typing import Any
 from typing import Dict
 from typing import Generator
+from typing import List
 from typing import Optional
 
 from autoblocks._impl import global_state
 from autoblocks._impl.config.constants import INGESTION_ENDPOINT
+from autoblocks._impl.testing.models import BaseEventEvaluator
+from autoblocks._impl.testing.models import Evaluation
+from autoblocks._impl.testing.models import TracerEvent
+from autoblocks._impl.util import SEND_EVENT_CORO_NAME
 from autoblocks._impl.util import AutoblocksEnvVar
+from autoblocks._impl.util import gather_with_max_concurrency
 
 log = logging.getLogger(__name__)
 
 
-@dataclass
+@atexit.register
+def _cleanup_tracer() -> None:
+    """
+    On program exit, attempt to flush any outstanding send event tasks.
+    Since the background thread is a daemon, we have to actively
+    check task counts and sleep to measure progress.
+    """
+    num_tries = 0
+
+    def get_pending() -> List[asyncio.Task[Any]]:
+        return [
+            task
+            for task in asyncio.all_tasks(loop=global_state.event_loop())
+            if task.get_coro().__name__ == SEND_EVENT_CORO_NAME and not task.done()  # type: ignore
+        ]
+
+    while len(get_pending()) > 0 and num_tries < 10:
```

Comment:
10 seconds is kinda short IMO, e.g. kubernetes default shutdown grace period is 30 seconds: https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-terminating-with-grace

Reply:
I think we can bump this up to 30 as that would be worse case 👍 In most scenarios I am hoping an immediate exit just happens (no events on queue)